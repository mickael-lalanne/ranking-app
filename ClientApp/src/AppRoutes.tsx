import Home from './components/Home';
import RankingLayout from './components/RankingLayout';
import { TEMPLATE_LAYOUT_PROPS, TIERLIST_LAYOUT_PROPS } from './models/RankingLayout';

export enum ERankingAppRoutes {
    templates = '/templates',
    tierlists = '/tierlists',
};

const AppRoutes = [
    {
        index: true,
        element: <Home />,
    },
    {
        path: ERankingAppRoutes.templates,
        element: <RankingLayout {...TEMPLATE_LAYOUT_PROPS} />,
    },
    {
        path: ERankingAppRoutes.tierlists,
        element: <RankingLayout {...TIERLIST_LAYOUT_PROPS} />,
    }
];

export default AppRoutes;
