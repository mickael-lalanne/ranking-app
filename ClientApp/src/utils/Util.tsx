import React from "react";

// All items that are not yet saved in the database starts with the following identifier
const TMP_ID_IDENTIFIER = '99999';

/**
 * Generate a temporary for items that are not yet saved in database
 * @returns {string} a random temporary id
 */
export const generateRandomId = (): string => {
    const randomNumber: number = Math.floor(Math.random() * 1000000000);
    return TMP_ID_IDENTIFIER + randomNumber;
};

/**
 * Check if an id is temporary
 * @param {string} id the id to check
 * @returns {boolean} true if the id is temporary
 */
export const isTemporaryId = (id: string): boolean => {
    return id.startsWith(TMP_ID_IDENTIFIER);
}

export interface ResizedImage {
    source: string; // the data URL of the image
    name: string;
}

/**
 * Resize an image while keeping its ratio
 * This function is called in the "onChange" event of input file
 * So we use the FileReader api to get the base image dimensions
 * @param {File} imageFile the image file to resize
 * @returns {Promise<ResizedImage>} an object containing the resized image as data URL and its name
 */
export const resizeImage = (imageFile: File): Promise<ResizedImage> => {
    const MAX_SIZE: number = 150;
    const reader = new FileReader();

    return new Promise ((resolve) => {
        reader.onload = () => {
            const tmpImg: HTMLImageElement = new Image();
    
            tmpImg.onload = () => {
                const resizeRatio = Math.min(MAX_SIZE / tmpImg.width, MAX_SIZE / tmpImg.height);

                // Initialize the canvas
                const canvas = document.createElement("canvas");
                const ctx = canvas.getContext('2d');
    
                // Set width and height
                canvas.width = tmpImg.width * resizeRatio;
                canvas.height = tmpImg.height * resizeRatio;
    
                // Draw image and export to a data-uri
                ctx!.drawImage(tmpImg, 0, 0, canvas.width, canvas.height);
    
                const dataURI = canvas.toDataURL();
                resolve({ source: dataURI, name: imageFile.name });
            };

            tmpImg.src = URL.createObjectURL(imageFile);
        }
        reader.readAsDataURL(imageFile);
    });

};

/**
 * @param array an array containing a createdAt field
 * @returns {Partial<{ createdAt: string }>[]} the same array but sorted by creation date
 */
export const sortByCreationDate = (array: Partial<{ createdAt: string }>[]) => {
    // Use slice() to avoid reference errors
    return array.slice().sort((a, b) => 
        new Date(a.createdAt!).getTime() - new Date(b.createdAt!).getTime()
    );
};

/**
 * Called in the template and tierlist editors
 * Returns a value for the MUI Tooltip component
 * @param {string?} content the text that should be displayed in the tooltip 
 * @returns {'' | React.JSX.Element | undefined} a span including the text if defined, otherwise undefined
 */
export const getTooltipTitleForSaveButtons = (
    content?: string
): '' | React.JSX.Element | undefined => {
    return content && <span style={{ fontSize: '15px'}}>{content}</span>;
}
