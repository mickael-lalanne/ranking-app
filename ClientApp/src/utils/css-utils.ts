export const ELEMENT_SIZE: string = '75px';
