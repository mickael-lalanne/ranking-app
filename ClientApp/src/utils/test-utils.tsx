import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from '../app/store';
import type { AppStore, RootState } from '../app/store';
import { ThemeOptions, ThemeProvider, createTheme } from '@mui/material/styles';
import '@testing-library/jest-dom';
import { DndProvider } from 'react-dnd';
import { TouchBackend } from 'react-dnd-touch-backend';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
export interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: Partial<RootState>;
    store?: AppStore;
}

// Mock MUI theme
const RANKING_APP_THEME_MOCK: ThemeOptions = {
    defaultRankingTheme: {
        primary: '',
        primaryLight: '',
        light: '',
        dark: '',
        info: '',
    },
};
const MockTheme = ({ children }: any) => {
    const theme = createTheme(RANKING_APP_THEME_MOCK);
    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = setupStore(preloadedState),
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
        return (
            <MockTheme>
                <Provider store={store}>
                    <DndProvider
                        backend={TouchBackend}
                        options={{ enableMouseEvents: true }}
                    >
                        {children}
                    </DndProvider>
                </Provider>
            </MockTheme>
        );
    }

    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
