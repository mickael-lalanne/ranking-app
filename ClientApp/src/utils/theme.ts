import { Theme, createTheme, alpha, ThemeOptions } from "@mui/material";

interface RankingTheme {
    primary: string;
    primaryLight: string;
    light: string;
    dark: string;
    info: string;
};
const PRIMARY_COLOR = '#BB2525';

declare module '@mui/material/styles' {
    interface Theme {
        defaultRankingTheme: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
    // allow configuration using `createTheme`
    interface ThemeOptions {
        defaultRankingTheme?: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
}

export const THEMES_COLORS: ThemeOptions['defaultRankingTheme'] = {
    primary: PRIMARY_COLOR,
    primaryLight: '#FF6969',
    light: '#FFF5E0',
    dark: '#141E46',
    info: '#9ED9EB'
}

export const RANKING_APP_THEME: Theme = createTheme({
    defaultRankingTheme: THEMES_COLORS,
    palette: {
        primary: {
            main: alpha(PRIMARY_COLOR, 0.7),
            light: alpha(PRIMARY_COLOR, 0.5),
            dark: alpha(PRIMARY_COLOR, 0.9),
        }
    }
});
