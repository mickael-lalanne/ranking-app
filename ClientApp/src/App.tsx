import React, { Component } from 'react';
import { Route, Routes } from 'react-router-dom';
import AppRoutes from './AppRoutes';
import Layout from './components/Layout';
import './custom.css';
import { ClerkProvider, SignedIn, SignedOut, RedirectToSignIn } from '@clerk/clerk-react';

if (!process.env.REACT_APP_CLERK_PUBLISHABLE_KEY) {
    throw new Error('Missing Publishable Key');
}
const clerkPubKey = process.env.REACT_APP_CLERK_PUBLISHABLE_KEY;

export default class App extends Component {
    static displayName = App.name;

    render() {
        return (
            <ClerkProvider publishableKey={clerkPubKey}>
                <Layout>
                    <SignedIn>
                        <Routes>
                            {AppRoutes.map((route, index) => {
                                const { element, ...rest } = route;
                                return (
                                    <Route key={index} {...rest} element={element} />
                                );
                            })}
                        </Routes>
                    </SignedIn>

                    <SignedOut>
                        <RedirectToSignIn />
                    </SignedOut>
                </Layout>
            </ClerkProvider>
        );
    }
}
