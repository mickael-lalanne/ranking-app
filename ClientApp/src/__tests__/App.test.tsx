import * as React from 'react';
import { screen } from '@testing-library/react';
import { renderWithProviders } from '../utils/test-utils';
import App from '../App';
import { MemoryRouter, Route, Router, Routes } from 'react-router-dom';
import { MemoryHistory, createMemoryHistory } from 'history';
import { ERankingAppRoutes } from '../AppRoutes';

const clerkPath: string = '@clerk/clerk-react';
jest.mock(clerkPath, () => ({
    ...jest.requireActual(clerkPath),
    useUser: jest.fn(() => {
        return { user: { fullName: 'Bruce Wayne' } };
    }),
    SignOutButton: () => <></>,
    SignedIn: ({ children }: React.PropsWithChildren<unknown>) => {
        return children;
    },
}));

const AppTest = (history: MemoryHistory) => {
    return (
        <Router location={history.location} navigator={history}>
            <App />
        </Router>
    );
};

describe('Screenshot', () => {
    it('takes default screenshot', async () => {
        const history = createMemoryHistory();
        history.push('/');
        const component = renderWithProviders(AppTest(history));

        expect(component.baseElement).toMatchSnapshot();
    });
});

describe('Common', () => {
    it('go to template page', async () => {
        const history = createMemoryHistory();
        history.push(ERankingAppRoutes.templates);

        renderWithProviders(AppTest(history));

        expect(screen.findByTestId('templates-viewer')).toBeDefined();
    });

    it('go to tierlist page', async () => {
        const history = createMemoryHistory();
        history.push(ERankingAppRoutes.tierlists);

        renderWithProviders(AppTest(history));

        expect(screen.findByTestId('tierlists-viewer')).toBeDefined();
    });
});
