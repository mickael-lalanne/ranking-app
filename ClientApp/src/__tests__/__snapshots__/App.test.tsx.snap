// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot takes default screenshot 1`] = `
<body>
  <div>
    <main
      class="css-irxpyf container"
      style="height: calc(100vh - 60px);"
    >
      <div
        class="css-jaw5y2"
      >
        <div
          class="css-13258ez"
        >
          <div
            class="css-1anxslm"
          >
            <h1>
              Hey Bruce Wayne
              <span>
                Welcome to your ranking app
              </span>
            </h1>
          </div>
        </div>
        <div
          class="css-l1dy4r"
        >
          <a
            class="css-1tq78vo nav-link"
            data-cy="template-button"
            href="/templates"
          >
            My templates
          </a>
          <a
            class="css-1tq78vo nav-link"
            data-cy="tierlist-button"
            href="/tierlists"
          >
            My Tierlists
          </a>
        </div>
        <div
          class="css-sp3apo"
        />
      </div>
    </main>
  </div>
  <script
    crossorigin="anonymous"
    data-clerk-publishable-key="pk_test_cmVuZXdlZC1ib2JjYXQtMi5jbGVyay5hY2NvdW50cy5kZXYk"
    src="https://renewed-bobcat-2.clerk.accounts.dev/npm/@clerk/clerk-js@4/dist/clerk.browser.js"
  />
</body>
`;
